### PR TITLE
Fix disabled text label rendering.

### PR DESCRIPTION
**Pull request by JeremyKuhne**
The problem was actually in ControlPaint.DrawStringDisabled that takes an IDeviceContext. In the big refactoring it lost the application of TextFormatFlags to decide whether or not to apply Graphics state. LinkLabel never set those flags historically so it didn't get the clipping it applied to the Graphics object applied to the HDC. Fix is to apply the flags properly.

I've also cleaned up the LinkLabel code and did a thorough pass of the TextExtensions to bring the code up to snuff.

Fixes https://github.com/dotnet/winforms/issues/7341

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8771)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs [hunk 1]
```diff
@@ -2040,7 +2040,7 @@ public static void DrawSizeGrip(Graphics graphics, Color backColor, int x, int y
             // This must come before creating the scope.
             FONT_QUALITY quality = TextRenderer.FontQualityFromTextRenderingHint(dc);
 
-            using var hdc = new DeviceContextHdcScope(dc);
+            using var hdc = new DeviceContextHdcScope(dc, TextRenderer.GetApplyStateFlags(dc, format));
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:2043]: This is the root of the bug.

File: src/System.Windows.Forms/src/System/Windows/Forms/LinkLabel.cs [hunk 2]
```diff
@@ -1302,13 +1094,12 @@ protected override void OnPaint(PaintEventArgs e)
 
                         PaintLinkBackground(g);
 
-                        // Because the code has been like that since long time, we assume that _textRegion is not null.
-                        g.IntersectClip(_textRegion!);
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/LinkLabel.cs:1306]: This is the only functional change, the rest is cleanup. I moved this inside of UseCompatibleTextRendering as it only really matters there.

**Comment by JeremyKuhne** [comment]
Fixes #7341

**Comment by elachlan** [comment]
> Fixes #7341

This will need to be added to the PR summary.

File: src/System.Windows.Forms/src/System/Windows/Forms/TextRenderer.cs [hunk 3]
```diff
@@ -8,640 +8,632 @@
 using System.Drawing;
 using System.Drawing.Text;
 
-namespace System.Windows.Forms
+namespace System.Windows.Forms;
+
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextRenderer.cs:12]: why this change?

**Review by dreddy-work — APPROVED** [review]
RLGTM.

File: src/System.Windows.Forms/src/System/Windows/Forms/TextRenderer.cs (hunk 3 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextRenderer.cs:12]: Reduces nesting, which improves maintainability

**Review by lonitra — APPROVED** [review]
🚢

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs [hunk 4]
```diff
@@ -2555,98 +2555,30 @@ private static ColorMatrix RemapBlackAndWhitePreserveTransparentMatrix(Color rep
...
-            }
-            else
-            {
-                result = TextFormatFlags.Left;
-            }
-
-            return result;
-        }
-
-        internal static StringFormat StringFormatForAlignment(ContentAlignment align)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:2642]: The removed functions were internal, are they not used anywhere?

**Review by Tanya-Solyanik — APPROVED** [review]
